path_oidc.go

```diff
@@ -0,0 +1,194 @@
+				logical.UpdateOperation: b.authURL,
+			},
+		},
+	}
+}
+
+// authURL returns a URL used for redirection to receive an authorization code.
+// This path requires a role name, or that a default_role has been configured.
+// Because this endpoint is unauthenticated, the response to invalid or non-OIDC
+// roles is intentionally non-descriptive and will simply be an empty string.
+//
+// TODO: I think we should probably just log all errors and return nil instead of
+//       the usual "return nil, err" pattern. We can still return a user error when
+//       they've not provided a required field though.
+func (b *jwtAuthBackend) authURL(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	authCodeURL := ""
+
+	config, err := b.config(ctx, req.Storage)
+	if err != nil {
+		return nil, err
+	}
+	if config == nil {
+		return logical.ErrorResponse("could not load configuration"), nil
+	}
+
+	roleName := d.Get("role").(string)
+	if roleName == "" {
+		roleName = config.DefaultRole
+		if roleName == "" {
+			return logical.ErrorResponse("missing role"), nil
+		}
+	}
+
+	role, err := b.role(ctx, req.Storage, roleName)
+	if err != nil {
+		return nil, err
+	}
+
+	redirectURI := d.Get("redirect_uri").(string)
+	if redirectURI == "" {
+		return logical.ErrorResponse("missing redirect_uri"), nil
+	}
+
+	if role != nil && role.RoleType == "oidc" {
+		provider, err := b.getProvider(ctx, config)
+		if err != nil {
+			return nil, errwrap.Wrapf("error getting provider for login operation: {{err}}", err)
+		}
+
+		if len(role.AllowedRedirectURIs) == 0 ||
+			strutil.StrListContains(role.AllowedRedirectURIs, redirectURI) {
+
+			// "openid" is a required scope for OpenID Connect flows
+			scopes := append([]string{oidc.ScopeOpenID}, role.OIDCScopes...)
+
+			// Configure an OpenID Connect aware OAuth2 client
+			oauth2Config := oauth2.Config{
+				ClientID:     config.OIDCClientID,
+				ClientSecret: config.OIDCClientSecret,
+				RedirectURL:  redirectURI,
+				Endpoint:     provider.Endpoint(),
+				Scopes:       scopes,
+			}
+
+			stateID, nonce, err := b.createState(roleName, redirectURI)
+			if err != nil {
+				return nil, errwrap.Wrapf("error generating OAuth state: {{err}}", err)
+			}
+
+			authCodeURL = oauth2Config.AuthCodeURL(stateID, oidc.Nonce(nonce))
+		}
+	}
+
+	resp := &logical.Response{
+		Data: map[string]interface{}{
+			"auth_url": authCodeURL,
+		},
+	}
+
+	return resp, nil
+}
+
+// createState make an expiring state object, associated with a random state ID
+// that is passed throughout the OAuth process. A nonce is also included in the
+// auth process, and for simplicity will be identical in length/format as the state ID.
+func (b *jwtAuthBackend) createState(rolename, redirectURI string) (string, string, error) {
+	randstr, err := base62.Random(2 * stateLength)
+	if err != nil {
+		return "", "", err
+	}
+
+	stateID, nonce := randstr[0:stateLength], randstr[stateLength:]
```

Comment:
Do we need to use base62 here? feels like go-uuid might be a faster method

Reply:
The base62 charset isn't required. I'll switch to go-uuid, but will just be for random bytes that will be printf'd into hex here, since the RFC calls for 160 bits and not 128.